# Neural Network – Data Preparation

## Objective
The goal of this notebook is to prepare a clean, scalable and comparable dataset
for Neural Network experiments.

We deliberately reuse the same processed dataset that was used by the winning
Gradient Boosted Decision Tree model to ensure:
- fair comparison
- consistent feature definitions
- reproducibility

This notebook focuses ONLY on data preparation.
Model architecture and training will be covered separately.



In [2]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.utils import shuffle

import warnings
warnings.filterwarnings("ignore")


In [3]:
DATA_PATH = "../data/processed/df_model.parquet"

df = pd.read_parquet(DATA_PATH)
print(df.shape)
df.head()



(2988177, 173)


,date_time,site_name,posa_continent,user_location_country,user_location_region,user_location_city,orig_destination_distance,user_id,is_mobile,is_package,channel,srch_ci,srch_co,srch_adults_cnt,srch_children_cnt,srch_rm_cnt,srch_destination_id,srch_destination_type_id,is_booking,cnt,hotel_continent,hotel_country,hotel_market,hotel_cluster,d1,d2,d3,d4,d5,d6,d7,d8,d9,d10,d11,d12,d13,d14,d15,d16,d17,d18,d19,d20,d21,d22,d23,d24,d25,d26,d27,d28,d29,d30,d31,d32,d33,d34,d35,d36,d37,d38,d39,d40,d41,d42,d43,d44,d45,d46,d47,d48,d49,d50,d51,d52,d53,d54,d55,d56,d57,d58,d59,d60,d61,d62,d63,d64,d65,d66,d67,d68,d69,d70,d71,d72,d73,d74,d75,d76,d77,d78,d79,d80,d81,d82,d83,d84,d85,d86,d87,d88,d89,d90,d91,d92,d93,d94,d95,d96,d97,d98,d99,d100,d101,d102,d103,d104,d105,d106,d107,d108,d109,d110,d111,d112,d113,d114,d115,d116,d117,d118,d119,d120,d121,d122,d123,d124,d125,d126,d127,d128,d129,d130,d131,d132,d133,d134,d135,d136,d137,d138,d139,d140,d141,d142,d143,d144,d145,d146,d147,d148,d149
0,2014-08-11 08:22:12,2,3,66,348,48862,2234.2641,12,0,1,9,2014-08-29,2014-09-02,2,0,1,8250,1,1,1,2,50,628,1,-2.283946,-2.346645,-2.400912,-2.367190,-2.273226,-1.722868,-2.392863,-2.381387,-2.239271,-2.259266,-2.305731,-2.400912,-2.369545,-1.712277,-2.399467,-1.901553,-2.065535,-2.384671,-2.399649,-1.140066,-2.354545,-2.400596,-2.400912,-2.365183,-2.397435,-2.365331,-2.397981,-2.363630,-1.675819,-2.194497,-2.400912,-2.399802,-2.398695,-2.392615,-2.209667,-2.317894,-1.281688,-2.398314,-2.374450,-2.370005,-2.031847,-2.387590,-1.645141,-2.400912,-2.378570,-2.364266,-2.397885,-2.355131,-1.349492,-2.373100,-2.397500,-2.348903,-2.241532,-2.400534,-2.207067,-2.305964,-2.397016,-2.241566,-2.393152,-2.399966,-2.380170,-2.391761,-2.399015,-2.400506,-2.400912,-2.363137,-2.395585,-2.396456,-2.391947,-2.400912,-2.356842,-2.379234,-2.393781,-1.967871,-1.960995,-2.399409,-2.333032,-2.385580,-2.257086,-2.316076,-2.391860,-2.397906,-2.372328,-2.079968,-2.220271,-2.400912,-2.399503,-2.387979,-2.392953,-2.400396,-1.938336,-2.396843,-1.366535,-2.381904,-2.396208,-2.306027,-2.399802,-2.398006,-2.385102,-2.400912,-2.387891,-2.340923,-2.359161,-2.336553,-2.364671,-2.396077,-2.280828,-1.760466,-2.371376,-2.389435,-2.368650,-2.392658,-2.397517,-2.356602,-2.399802,-2.397591,-2.392110,-2.394294,-2.398695,-2.392041,-1.925547,-1.868743,-2.400912,-2.389131,-2.390027,-2.363872,-2.396035,-2.007101,-2.400837,-2.120344,-2.275190,-2.203799,-2.373795,-2.365093,-2.400912,-2.370760,-2.343093,-2.320423,-1.732640,-2.384553,-2.345528,-2.396591,-2.399953,-2.388116,-2.394294,-2.400667,-2.398716,-2.386585,-2.390370
1,2014-02-27 18:01:32,2,3,66,318,52078,NaN,756,0,1,4,2014-04-18,2014-04-20,2,0,1,8291,1,1,1,2,50,191,2,-2.292252,-2.299437,-2.299516,-2.081118,-2.012726,-1.846311,-2.256943,-2.241929,-2.296619,-2.272894,-2.089446,-2.299516,-2.073694,-2.126230,-2.299516,-1.907082,-2.224969,-2.130352,-2.299516,-2.299512,-2.286226,-2.294937,-2.299516,-2.248783,-2.256279,-2.271063,-2.077581,-2.278924,-2.019199,-2.170334,-2.299516,-2.299516,-2.298682,-2.134491,-2.181788,-2.298920,-1.510314,-2.299516,-2.299516,-2.272754,-2.122334,-2.264607,-1.861097,-2.299510,-2.249796,-2.259944,-2.292342,-2.169161,-1.785864,-2.284237,-2.293556,-2.288045,-2.050526,-2.299516,-2.192732,-2.055975,-2.287063,-1.835956,-2.271718,-2.299516,-2.283584,-2.299516,-2.299516,-2.271189,-2.299516,-2.166047,-2.265805,-2.290143,-2.288796,-2.299516,-2.276115,-2.291246,-2.165692,-2.092732,-2.121190,-2.298750,-2.267115,-1.875826,-2.231525,-1.893010,-2.262565,-2.159382,-2.261873,-2.204839,-2.276740,-2.299516,-2.299439,-2.268711,-2.297019,-2.299516,-2.046545,-2.299516,-1.561541,-2.083417,-2.299516,-2.251615,-2.167010,-2.299516,-2.286313,-2.299516,-2.218038,-2.258407,-2.269182,-2.294434,-2.291297,-2.288771,-2.299516,-1.880610,-2.172632,-2.298266,-2.297850,-2.250530,-2.299516,-2.288303,-2.299516,-2.299516,-2.298682,-2.299516,-2.298682,-2.282234,-2.216793,-1.924212,-2.299516,-2.298510,-2.299077,-2.241868,-2.298671,-2.135436,-2.242227,-1.986876,-2.297201,-1.858242,-2

The dataset already includes:
- engineered numerical features
- encoded categorical variables
- joined destination latent features


In [4]:
TARGET_COL = "hotel_cluster"

X = df.drop(columns=[TARGET_COL])
y = df[TARGET_COL]


In [5]:
MAX_SAMPLES = 500_000  # adjustable

if len(X) > MAX_SAMPLES:
    X, y = shuffle(X, y, random_state=42)
    X = X.iloc[:MAX_SAMPLES]
    y = y.iloc[:MAX_SAMPLES]

print(X.shape, y.shape)


(500000, 172) (500000,)


Neural Networks scale poorly on large tabular datasets without GPU acceleration.
For this educational project we intentionally limit the dataset size to keep
training time reasonable (<10–15 minutes).


In [6]:
X_train, X_val, y_train, y_val = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

print(X_train.shape, X_val.shape)


(400000, 172) (100000, 172)


In [7]:
scaler = StandardScaler()

X_train_scaled = scaler.fit_transform(X_train)
X_val_scaled = scaler.transform(X_val)


ValueError: could not convert string to float: '2014-08-05 08:16:53'

Neural Networks are sensitive to feature scales.
Standardization (zero mean, unit variance) is mandatory to ensure:
- stable gradients
- faster convergence
- fair contribution of all features


In [8]:
X_train.dtypes[X_train.dtypes == "object"]


date_time    object
srch_ci      object
srch_co      object
dtype: object

In [9]:
X_train.dtypes


date_time                 object
site_name                  int64
posa_continent             int64
user_location_country      int64
user_location_region       int64
                          ...   
d145                     float32
d146                     float32
d147                     float32
d148                     float32
d149                     float32
Length: 172, dtype: object

In [10]:
def parse_datetime_features(df):
    df = df.copy()
    
    # Convert to datetime
    df["date_time"] = pd.to_datetime(df["date_time"], errors="coerce")
    df["srch_ci"] = pd.to_datetime(df["srch_ci"], errors="coerce")
    df["srch_co"] = pd.to_datetime(df["srch_co"], errors="coerce")
    
    # Date time features
    df["search_year"] = df["date_time"].dt.year
    df["search_month"] = df["date_time"].dt.month
    df["search_day"] = df["date_time"].dt.day
    df["search_dow"] = df["date_time"].dt.dayofweek
    
    # Stay duration
    df["stay_length"] = (df["srch_co"] - df["srch_ci"]).dt.days
    
    # Drop original datetime columns
    df = df.drop(columns=["date_time", "srch_ci", "srch_co"])
    
    return df


In [11]:
X_train = parse_datetime_features(X_train)
X_val = parse_datetime_features(X_val)


In [12]:
X_train = X_train.fillna(0)
X_val = X_val.fillna(0)


In [13]:
scaler = StandardScaler()

X_train_scaled = scaler.fit_transform(X_train)
X_val_scaled = scaler.transform(X_val)


In [14]:
X_train_scaled.shape, X_val_scaled.shape


((400000, 174), (100000, 174))

In [15]:
np.mean(X_train_scaled), np.std(X_train_scaled)


(8.25799148092484e-16, 0.9971222961920057)

## Neural Network Model

After completing data preprocessing, feature engineering, and scaling, the dataset is now fully suitable for training a neural network model.

All input features are numerical, dense, and standardized, which is a strict requirement for neural network architectures. Datetime features were transformed into meaningful numerical representations (e.g. year, month, day, length of stay), and missing values were handled using simple imputation.

In this section, we train a baseline Multi-Layer Perceptron (MLP) classifier and compare its performance to the previously trained Gradient Boosted Decision Tree model, which served as our winning baseline.

The goal of this experiment is not to outperform the tree-based model at all costs, but to:
- evaluate whether a neural network can learn useful patterns from the same feature set,
- compare training stability and computational cost,
- and assess model suitability for this type of structured tabular data.


In [16]:
import torch
import torch.nn as nn
import torch.optim as optim
from torch.utils.data import TensorDataset, DataLoader


In [17]:
X_train_tensor = torch.tensor(X_train_scaled, dtype=torch.float32)
y_train_tensor = torch.tensor(y_train.values, dtype=torch.long)

X_val_tensor = torch.tensor(X_val_scaled, dtype=torch.float32)
y_val_tensor = torch.tensor(y_val.values, dtype=torch.long)


In [18]:
BATCH_SIZE = 1024

train_loader = DataLoader(
    TensorDataset(X_train_tensor, y_train_tensor),
    batch_size=BATCH_SIZE,
    shuffle=True
)

val_loader = DataLoader(
    TensorDataset(X_val_tensor, y_val_tensor),
    batch_size=BATCH_SIZE,
    shuffle=False
)


In [19]:
class HotelClusterNN(nn.Module):
    def __init__(self, input_dim, num_classes):
        super().__init__()
        
        self.net = nn.Sequential(
            nn.Linear(input_dim, 256),
            nn.ReLU(),
            nn.BatchNorm1d(256),
            nn.Dropout(0.3),
            
            nn.Linear(256, 128),
            nn.ReLU(),
            nn.BatchNorm1d(128),
            nn.Dropout(0.3),
            
            nn.Linear(128, num_classes)
        )
    
    def forward(self, x):
        return self.net(x)


In [20]:
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

model = HotelClusterNN(
    input_dim=X_train_scaled.shape[1],
    num_classes=y_train.nunique()
).to(device)

criterion = nn.CrossEntropyLoss()
optimizer = optim.Adam(model.parameters(), lr=1e-3)


In [21]:
EPOCHS = 5

for epoch in range(EPOCHS):
    model.train()
    train_loss = 0
    
    for X_batch, y_batch in train_loader:
        X_batch = X_batch.to(device)
        y_batch = y_batch.to(device)
        
        optimizer.zero_grad()
        outputs = model(X_batch)
        loss = criterion(outputs, y_batch)
        loss.backward()
        optimizer.step()
        
        train_loss += loss.item()
    
    train_loss /= len(train_loader)
    
    print(f"Epoch {epoch+1}/{EPOCHS} - Train loss: {train_loss:.4f}")


Epoch 1/5 - Train loss: 3.6696
Epoch 2/5 - Train loss: 3.4480
Epoch 3/5 - Train loss: 3.3979
Epoch 4/5 - Train loss: 3.3672
Epoch 5/5 - Train loss: 3.3462


In [22]:
def top_k_accuracy(model, dataloader, k=5):
    model.eval()
    correct = 0
    total = 0
    
    with torch.no_grad():
        for X_batch, y_batch in dataloader:
            X_batch = X_batch.to(device)
            y_batch = y_batch.to(device)
            
            outputs = model(X_batch)
            _, top_k_preds = outputs.topk(k, dim=1)
            
            correct += (top_k_preds == y_batch.unsqueeze(1)).any(dim=1).sum().item()
            total += y_batch.size(0)
    
    return correct / total


In [23]:
top5_acc = top_k_accuracy(model, val_loader, k=5)
top5_acc


0.45447

## Neural Network Results and Comparison

The neural network was trained using a standard multi-class classification setup with CrossEntropyLoss and evaluated using Top-5 Accuracy on the validation set.

### Training Behavior

The training loss decreased steadily across epochs:

- Epoch 1: 3.67
- Epoch 5: 3.35

This indicates stable learning, proper feature scaling, and no signs of divergence or overfitting.

### Validation Performance

- Top-5 Accuracy: **0.454**

This means that in approximately 45% of validation cases, the correct hotel cluster appeared within the top 5 predictions of the neural network.

### Comparison with Gradient Boosted Decision Tree

| Model | Metric | Score |
|------|------|------|
| LightGBM (GBDT) | Hit@5 | **0.535** |
| LightGBM (GBDT) | MAP@5 | **0.311** |
| Neural Network (MLP) | Top-5 Accuracy | **0.454** |

While the Gradient Boosted Decision Tree outperforms the neural network, the gap is moderate rather than dramatic. This outcome is expected, as tree-based models are known to perform exceptionally well on structured tabular data with rich categorical features.

Additionally, the GBDT model was trained using a ranking-based objective aligned directly with the competition metric (MAP@5), whereas the neural network was trained as a standard classifier.

### Conclusion

The neural network provides a solid baseline and demonstrates that meaningful signal can be learned from the same feature set. However, for this problem, the Gradient Boosted Decision Tree remains the preferred production model due to its superior performance, faster training, and better alignment with the evaluation metric.

